src/Spice86.Core/Emulator/Devices/Sound/PCSpeaker/PcSpeaker.cs

```diff
@@ -1,221 +1,301 @@
+        
+        _deviceThread = new DeviceThread(nameof(PcSpeaker), PlaybackLoop, pauseHandler, loggerService);
     }
-
-    /// <summary>
-    /// Invoked when the frequency has changed.
-    /// </summary>
-    private void OnTimerSettingChanged() {
-        if (_currentPeriod == PeriodInSamples) {
+    
+    private void OnPitSettingChanged(object? sender, EventArgs e) {
+        // Only respond to changes in channel 2 (PC Speaker)
+        if (_pit8254Counter.Index != 2) {
             return;
         }
-        if (_loggerService.IsEnabled(Serilog.Events.LogEventLevel.Verbose)) {
-            _loggerService.Verbose("PC Speaker frequency changed to {Frequency}", _pit8254Counter.Frequency);
+        
+        if (_prevPitMode != _pit8254Counter.Mode) {
+            SetPITControl((PitMode)_pit8254Counter.Mode);
+            _prevPitMode = _pit8254Counter.Mode;
         }
-        EnqueueCurrentNote();
-
-        _durationTimer.Reset();
-        _durationTimer.Start();
-        _currentPeriod = PeriodInSamples;
-        _deviceThread.Resume();
+        
+        SetCounter(_pit8254Counter.ReloadValue, _pit8254Counter.Mode);
     }
-
-    /// <summary>
-    /// Enqueues the current note.
-    /// </summary>
-    private void EnqueueCurrentNote() {
-        if (!_durationTimer.IsRunning || _currentPeriod == 0) {
+    
+    private void PlaybackLoop() {
+        Array.Clear(_audioBuffer, 0, _audioBuffer.Length);
+        GenerateAudio();
+        ApplyFiltersAndRender();
+    }
+    
+    private void GenerateAudio() {
+        if (!_portB.SpeakerOutput || !_portB.Timer2Gating || !_pitState.Mode3Counting) {
+            // If speaker is disabled or gate is off, generate silence
+            Array.Fill(_audioBuffer, 0.0f);
             return;
         }
-        _durationTimer.Stop();
-
-        int periodDuration = _ticksPerSample * _currentPeriod;
-        int repetitions = (int)(_durationTimer.ElapsedTicks / periodDuration);
-        _currentNote = new QueuedNote(_currentPeriod, repetitions);
+        
+        if (_pitState.Mode == PitMode.SquareWave || _pitState.Mode == PitMode.SquareWaveAlias) {
+            GenerateSquareWave();
+        } else {
+            // For all other modes or inactive state, just use the current amplitude
+            Array.Fill(_audioBuffer, _pitState.Amplitude);
+        }
     }
-
-    /// <summary>
-    /// Fills a buffer with a square wave of the current frequency.
-    /// </summary>
-    /// <param name="buffer">Buffer to fill.</param>
-    /// <param name="period">The number of samples in the period.</param>
-    /// <returns>Number of bytes written to the buffer.</returns>
-    private int GenerateSquareWave(Span<byte> buffer, int period) {
-        if (period < 2) {
-            buffer[0] = 127;
-            return 1;
+    
+    private void GenerateSquareWave() {
+        // Calculate frequency from PIT counter value
+        float counterMs = MsPerPitTick * _pit8254Counter.ReloadValue;
```

Comment:
Shouldn't the pit be responsible for computing this?

Reply:
Yes, I'll move it in the PIT.